### PR TITLE
librados: use CEPH_OSD_FLAG_FULL_FORCE for IoCtxImpl::remove

### PR DESCRIPTION
**Pull request by chenyuanrun**
librados::OPERATION_FULL_FORCE should be translated to CEPH_OSD_FLAG_FULL_FORCE before calling IoCtxImpl::remove().

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by chenyuanrun** [comment]
@ajarr @gregsfortytwo I think this is a BUG for librados, could you please review this fix? Thanks!

**Comment by chenyuanrun** [comment]
@tchaikov fix commit https://github.com/ceph/ceph/commit/435954ba6073c51e71117ee1a4ce16ffb71eb301

**Comment by charmingcow** [comment]
It seems like only at the delete blocking ABIs, against aio ABIs is ok.
Could help to review this pr?
@xiexingguo  @idryomov

**Review by idryomov — COMMENTED** [review]
The change seems correct, but doesn't this issue also affect
```
int librados::IoCtx::remove(const std::string& oid, int flags)
{
  object_t obj(oid);
  return io_ctx_impl->remove(obj, flags); 
}
```
Here flags are passed from a public API to a private implementation
```
int librados::IoCtxImpl::remove(const object_t& oid, int flags)
{
  ::ObjectOperation op;
  prepare_assert_ops(&op);
  op.remove();
  return operate(oid, &op, NULL, flags);
}
```
without any translation.

It would be nice to:
- file a ticket on https://tracker.ceph.com ticket so the fix can be backported
- add an integration test to ensure that this doesn't break in the future

**Comment by chenyuanrun** [comment]
@idryomov Thanks! I will file a ticket on https://tracker.ceph.com/ later.

**Comment by chenyuanrun** [comment]
@idryomov `librados::IoCtxImpl::remove` can not be fixed because it is used by src/tools/rados/rados.cc with the private flags:

https://github.com/ceph/ceph/blob/ed736fab6e085a92a26ee35dbee96d20fcd1b10b/src/tools/rados/rados.cc#L3077-L3083

If we change it , we break the public API, and the break the apps which rely on the improper behavior of this API.

**Comment by idryomov** [comment]
> If we change it , we break the public API, and the break the apps which rely on the improper behavior of this API.

We need to at least document this -- I don't see anything (not even a one-liner comment!) in `src/include/rados/librados.hpp` or `src/librados/librados_cxx.cc`.  It's very easy to get confused as it is.

Then this is up to @ceph/core, but I'd also suggest deprecating `librados::IoCtx::remove(const std::string& oid, int flags)`.  We could add `librados::IoCtx::remove2()` taking the right flags as a replacement or perhaps just change `librados::IoCtx::remove(const std::string& oid)` to pass both `FULL_TRY` and `FULL_FORCE` as done in `src/tools/rados/rados.cc` in the snippet that you highlighted in the previous comment.  That would effectively make `--force-full` behavior the default which seems like what @jdurgin had in mind in https://tracker.ceph.com/issues/22413#note-2.

**Comment by idryomov** [comment]
> or perhaps just change `librados::IoCtx::remove(const std::string& oid)` to pass both `FULL_TRY` and `FULL_FORCE` as done in `src/tools/rados/rados.cc` in the snippet that you highlighted in the previous comment. That would effectively make `--force-full` behavior the default which seems like what @jdurgin had in mind in https://tracker.ceph.com/issues/22413#note-2.

Note from git history spelunking: at the time when https://github.com/ceph/ceph/pull/20534 landed, only `FULL_FORCE` was passed in `src/tools/rados/rados.cc`.  `FULL_TRY` was added later in https://github.com/ceph/ceph/pull/24264.

File: src/librados/IoCtxImpl.cc [hunk 1]
```diff
@@ -1235,7 +1235,7 @@ int librados::IoCtxImpl::remove(const object_t& oid)
   ::ObjectOperation op;
   prepare_assert_ops(&op);
   op.remove();
-  return operate(oid, &op, nullptr, librados::OPERATION_FULL_FORCE);
+  return operate(oid, &op, nullptr, CEPH_OSD_FLAG_FULL_FORCE);
```
Comment (rzarzynski) [inline src/librados/IoCtxImpl.cc:1238]: Let's take a look on the flow:

```cpp
int librados::IoCtxImpl::operate(const object_t& oid, ::ObjectOperation *o,
                                 ceph::real_time *pmtime, int flags)
{
  // ...
  Objecter::Op *objecter_op = objecter->prepare_mutate_op(
    oid, oloc,
    *o, snapc, ut,
    flags | extra_op_flags,
    oncommit, &ver);
```

```cpp
  // mid-level helpers
  Op *prepare_mutate_op(
    const object_t& oid, const object_locator_t& oloc,
    ObjectOperation& op, const SnapContext& snapc,
    ceph::real_time mtime, int flags,
    Context *oncommit, version_t *objver = NULL,
    osd_reqid_t reqid = osd_reqid_t(),
    ZTracer::Trace *parent_trace = nullptr) {
    Op *o = new Op(oid, oloc, std::move(op.ops), flags | global_op_flags |
                   CEPH_OSD_FLAG_WRITE, oncommit, objver,
                   nullptr, parent_trace);
```

The union with `CEPH_OSD_FLAG_WRITE` suggests the conceptual type.

All constructors of `Op` do:

```cpp
  struct Op : public RefCountedObject {
    // ...
    op_target_t target;
    // ...
    Op(const object_t& o, const object_locator_t& ol,  osdc_opvec&& _ops,
       int f, /* ... */) :
      target(o, ol, f),
```

```cpp
  struct op_target_t {
    int flags = 0;
    // ...
    op_target_t(const object_t& oid, const object_locator_t& oloc, int flags)
      : flags(flags),
    // ...
    bool respects_full() const {
      return
        (flags & (CEPH_OSD_FLAG_WRITE | CEPH_OSD_FLAG_RWORDERED)) &&
        !(flags & (CEPH_OSD_FLAG_FULL_TRY | CEPH_OSD_FLAG_FULL_FORCE));
    }
```

`respects_full()` clearly assumes `CEPH_OSD_FLAG_FULL_FORCE`.

**Comment by chenyuanrun** [comment]
> Note from git history spelunking: at the time when https://github.com/ceph/ceph/pull/20534 landed, only FULL_FORCE was passed in src/tools/rados/rados.cc. FULL_TRY was added later in https://github.com/ceph/ceph/pull/24264.

@idryomov Should I change CEPH_OSD_FLAG_FULL_FORCE to CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY so that this fix could be merged?

**Comment by idryomov** [comment]
That would be my preference, as @jdurgin clearly wanted `remove()` to behave like that CLI command:
> You can get around this by using rados_write_op_operate with the 'LIBRADOS_OPERATION_FULL_FORCE' flag (128), like the rados cli, but I agree this should be the default for plain old remove().

But the current fix makes sense on its own, so one argue that `FULL_FORCE` -> `FULL_FORCE | FULL_TRY` should be a separate change.  @rzarzynski would have the final say here.

**Comment by rzarzynski** [comment]
These flags have rich history. `FULL_TRY` seems to be even stronger than `FULL_FORCE`:

```cpp
  // discard due to cluster full transition?  (we discard any op that
  // originates before the cluster or pool is marked full; the client
  // will resend after the full flag is removed or if they expect the
  // op to succeed despite being full).  The except is FULL_FORCE and
  // FULL_TRY ops, which there is no reason to discard because they
  // bypass all full checks anyway.  If this op isn't write or
  // read-ordered, we skip.
  // FIXME: we exclude mds writes for now.
  if (write_ordered && !(m->get_source().is_mds() ||
                         m->has_flag(CEPH_OSD_FLAG_FULL_TRY) ||
                         m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) &&
      info.history.last_epoch_marked_full > m->get_map_epoch()) {
    dout(10) << __func__ << " discarding op sent before full " << m << " "
             << *m << dendl;
    return;
  }
  // mds should have stopped writing before this point.
  // We can't allow OSD to become non-startable even if mds
  // could be writing as part of file removals.
  if (write_ordered && osd->check_failsafe_full(get_dpp()) &&
      !m->has_flag(CEPH_OSD_FLAG_FULL_TRY)) {
    dout(10) << __func__ << " fail-safe full check failed, dropping request." << dendl;
    return;
  }
```

The failsafe check has been relaxed in https://github.com/ceph/ceph/pull/17177. The idea was to accept operations that are _net reducers of usage_. However, I'm not sure it can be said with certainty  about a deletion of an RADOS object. To exemplify, let's imagine removing 1 byte long obj. Is it a _net reducer_? Well, on one hand it going to squeeze at least 1 bytes from an `ObjectStore` (how much exactly is an implementation's detail); on the other it will result in generation of a few dozens of bytes in `PGLog` and, maybe, trimming other entries.

Pretty complex thing. I would prefer to be cautious when it comes to the failsafe check.

File: src/librados/IoCtxImpl.cc (hunk 1 above)
Comment (rzarzynski) [reply src/librados/IoCtxImpl.cc:1238]: The `respects_full()` governs the objecter's behavior on traffic pausing:

```cpp
bool Objecter::target_should_be_paused(op_target_t *t)
{
  const pg_pool_t *pi = osdmap->get_pg_pool(t->base_oloc.pool);
  bool pauserd = osdmap->test_flag(CEPH_OSDMAP_PAUSERD);
  bool pausewr = osdmap->test_flag(CEPH_OSDMAP_PAUSEWR) ||
    (t->respects_full() && (_osdmap_full_flag() || _osdmap_pool_full(*pi)));

  return (t->flags & CEPH_OSD_FLAG_READ && pauserd) ||
    (t->flags & CEPH_OSD_FLAG_WRITE && pausewr) ||
    (osdmap->get_epoch() < epoch_barrier);
}
```

However, the main responsibility of `op_target_t::flags` is being passed to an OSD via `MOSDOp`:

```cpp
Objecter::MOSDOp *Objecter::_prepare_osd_op(Op *op)
{
  // rwlock is locked

  int flags = op->target.flags;
  flags |= CEPH_OSD_FLAG_KNOWN_REDIR;
  flags |= CEPH_OSD_FLAG_SUPPORTSPOOLEIO;

  // Nothing checks this any longer, but needed for compatibility with
  // pre-luminous osds
  flags |= CEPH_OSD_FLAG_ONDISK;

  if (!honor_pool_full)
    flags |= CEPH_OSD_FLAG_FULL_FORCE;

  // ...
  auto m = new MOSDOp(client_inc, op->tid,
                      hobj, op->target.actual_pgid,
                      osdmap->get_epoch(),
                      flags, op->features);
```

**Comment by rzarzynski** [comment]
> If we change it , we break the public API, and the break the apps which rely on the improper behavior of this API.

I agree with @idryomov on the doc and deprecating the private-flags-taking variant of `remove()`. We should get `remove2()` doing the proper translation (see `translate_flags()` in `librados/librados_util.cc`).

**Comment by idryomov** [comment]
@rzarzynski So I take it that you prefer to see a new `remove2()` and have plain old `remove()` fixed to pass the proper flag only for `FULL_FORCE` and not `FULL_TRY`?  I'm not sure I see much use in that...

It might be better to have plain old `remove()` not pass any flags then -- this would avoid confusion as passing `FULL_FORCE` is only a half measure and also wouldn't be a change in behavior compared to today since the public `FULL_FORCE` that is passed today is clearly bogus.

**Comment by idryomov** [comment]
> To exemplify, let's imagine removing 1 byte long obj. Is it a _net reducer_? Well, on one hand it going to squeeze at least 1 bytes from an `ObjectStore` (how much exactly is an implementation's detail); on the other it will result in generation of a few dozens of bytes in `PGLog` and, maybe, trimming other entries.

I thought the way `FULL_TRY` was supposed to work is that it's always safe to pass and is effectively free, because if the op isn't a net reducer (let's define that as "makes the situation worse for the OSD"), it's failed with `ENOSPC` error.  Am I mistaken?

**Comment by rzarzynski** [comment]
> @rzarzynski So I take it that you prefer to see a new `remove2()` and have plain old `remove()` fixed to pass the proper flag only for `FULL_FORCE` and not `FULL_TRY`? I'm not sure I see much use in that...

Yes, just `FULL_FORCE`, without `FULL_TRY` (I know, the naming became confusing).

The rationale would be to let `removes()` squeeze some data when it's still quite safe – under the _full_ condition but stay really cautious (paranoid?) when things had gone so badly that OSD got into _failsafe full_.

> I thought the way `FULL_TRY` was supposed to work was that it's always safe to pass and is effectively free, because if the op isn't a net reducer (let's define that as "makes the situation worse for the OSD"), it's failed with `ENOSPC` error. 

I think you means the second check which happens later, in `prepare_transaction()`, and bases on the `delta_stats`:

```cpp
  // check for full
  if ((ctx->delta_stats.num_bytes > 0 ||
       ctx->delta_stats.num_objects > 0) &&  // FIXME: keys?
      pool.info.has_flag(pg_pool_t::FLAG_FULL)) {
    auto m = ctx->op->get_req<MOSDOp>();
    if (ctx->reqid.name.is_mds() ||   // FIXME: ignore MDS for now
        m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) {
      dout(20) << __func__ << " full, but proceeding due to FULL_FORCE or MDS"
               << dendl;
    } else if (m->has_flag(CEPH_OSD_FLAG_FULL_TRY)) {
      // they tried, they failed.
      dout(20) << __func__ << " full, replying to FULL_TRY op" << dendl;
      return pool.info.has_flag(pg_pool_t::FLAG_FULL_QUOTA) ? -EDQUOT : -ENOSPC;
    } else {
      // drop request
      dout(20) << __func__ << " full, dropping request (bad client)" << dendl;
      return -EAGAIN;
    }
  }
```

I doubt these counters cover size of `PGLog` entries, internals of an `ObjectStore`' and so on. I perceive the _failsafe full_ as a measure to (imperfectly) cover such circumstances and prevent an OSD from becoming unable to boot. I imagine it can be very hard for `ObjectStore` to be always exact about `ENOSPC` – I'm thinking here about things like a RocksDB's compaction-on-startup in `BlueStore`.

**Comment by chenyuanrun** [comment]
@idryomov @rzarzynski So what should I do if I want this fix could be merged?

**Review by idryomov — APPROVED** [review]
No objections from me to this going in as is.

**Comment by idryomov** [comment]
> @idryomov @rzarzynski So what should I do if I want this fix could be merged?

Since this is a bug fix, I think we need a tracker ticket filed for backporting purposes.

**Comment by chenyuanrun** [comment]
@idryomov I have created a issue: https://tracker.ceph.com/issues/64558

**Comment by idryomov** [comment]
> @idryomov I have created a issue: https://tracker.ceph.com/issues/64558

Now need to add `Fixes: https://tracker.ceph.com/issues/64558` on the line before `Signed-off-by` in the commit message.  See https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#fixes-lines for details.

**Review by rzarzynski — APPROVED** [review]

**Comment by chenyuanrun** [comment]
@idryomov I have modified my commit message, could you please review it again? Thanks!

**Comment by amathuria** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues64974

**Pull request merged**